### PR TITLE
Basic functionality for moving hand and reading hand state data and finger sensor data

### PR DESCRIPTION
**Pull request by palanc**
Basic functionality for moving hand and reading hand state data and finger sensor data

### CODE REVIEW

**Comment by sniyaz** [comment]
I took a look through, and there's one big design decision I want to clarify (I'm not too great with low-level ROS stuff, so sorry if I say something dumb 😅)

Do we want to:

1) Use `pr_hardware_interfaces` the same way that `barrett_ros_control` does (like [here](https://github.com/personalrobotics/barrett_ros_control/blob/2d66afc06fe51329050c00b436bf256d6a6cd44d/src/hardware/HandHW.cpp))? If so, it looks like there's a lot of refactoring and re-writing of code required to do this.

2) Or just keep the API that's been defined here for reading the state of the gripper and sending commands? In this case there would be more hacking required in `libherb`/`AIKIDO` that doesn't follow the convention we use for the Barrett hand (but it's a shorter path to getting something working).

If we want to go with `2`, I'll go ahead and request some nits. If we want to go with `1` we should touch base with people who are more knowledgable about how our ROS controllers work. We can also definitely discuss at meeting tomorrow 😄

**Comment by palanc** [comment]
I agree, we should discuss with others on the best way to integrate into AIKIDO.

**Comment by brianhou** [comment]
I've been looking into this for the neck as well. I think there's a middle ground where we can use `pr_hardware_interfaces` without having to set up the complicated control loop that `barrett_ros_control` uses. Let's talk offline.

File: wsg_50_driver/include/wsg_50/fmf_weiss_finger.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * @brief The FMFWeissFinger class
+*/
+class FMFWeissFinger: public WeissFinger {
+public:
+    /**
+     * @brief FMFWeissFinger: Constructor
+     * @param nh: Node handle for grabbing params
+     * @param param_prefix: Prefix for param keys
+    */
+    FMFWeissFinger(ros::NodeHandle& nh, std::string param_prefix);
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/fmf_weiss_finger.h]: We don't need to pass these params through a ROS node, right? Can we change this to take some struct of params and leave the ROS stuff for another library?

File: wsg_50_driver/include/wsg_50/hand.h [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    // Threshold for determining if the hand is moving or not
+    static constexpr double NOT_MOVING_SPEED_THRESH = 0.01;
+
+    /**
+     * @brief Hand: Constructor
+     * @param nh: Node handle for grabbing params
+     * @param param_prefix: Prefix for param keys
+    */
+    Hand(ros::NodeHandle& nh, std::string param_prefix);
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/hand.h]: Ditto, can we pass a param struct or something?

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+public:
+    // Command indicating transmission/reception of data to/from hand
+    static const unsigned char FINGER_CMD_ID = 0xB3;
+
+    /**
+     * @brief OpticalWeissFinger: Constructor
+     * @param nh: Node handle for grabbing params
+     * @param param_prefix: Prefix for param keys
+    */
+    OpticalWeissFinger(ros::NodeHandle& nh, std::string param_prefix);
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/optical_weiss_finger.h]: Ditto, param struct?

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h [hunk 4]
```diff
@@ -0,0 +1,341 @@
+#ifndef OPTICALWEISSFINGER_H
+#define OPTICALWEISSFINGER_H
+
+#include <stdlib.h>
+#include <vector>
+#include <mutex>
+#include <thread>
+#include <deque>
+#include "ros/ros.h"
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/optical_weiss_finger.h]: We should remove all of these `ros` includes

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    /**
+     * @brief hand_width_to_finger_tip_separation: Computes the distance from this finger to the target on
+     *        the other finger.
+     * @param width: The current width of the hand
+     * @return The distance between the surface of this finger and the target on the other fingertip
+    */
+    double hand_width_to_finger_tip_separation(double width);
+private:
+
+    ros::NodeHandle nh_; // Node handle used to create publisher of data
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/optical_weiss_finger.h]: We should strip this out- I think we want to leave all the `ros` stuff for another library.

Comment (sniyaz) [reply wsg_50_driver/include/wsg_50/optical_weiss_finger.h]: Another thing too is that we should have the data originally published by `ros` nodes here easily accessible from methods that whatever `ros` layer we write can call.

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    bool initialized_; // Status flag, indicates if object has been initialized
+    bool can_calibrate_; // Status flag, indicates if object is able to perform calibration
+    std::deque<wsg_50_common::WeissFingerData> data_buff_; // Contains the most recent data samples
+    std::mutex data_buffer_mutex_; // Mutex that controls access to data_buff_
+    std::thread* read_finger_thread_; // Thread for reading data from the finger
+    bool read_finger_alive_; // Once thread has started, will shutdown if this is false
+    bool paused_; // Indicates if data is being read from the finger
+    std::string pub_topic_; // Topic to publish data to
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/optical_weiss_finger.h]: This should disappear.

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    unsigned int devices_; // The number of optical sensors that this finger contains
+    std::vector<uint16_t> reg_addrs_; // Addresses of the optical sensor registers to read
+    std::vector<uint8_t> reg_lengths_; // Lengths of the optical sensor registers to read
+    double finger_read_rate_; // The rate at which data from the finger should be read
+    int data_buff_max_size_; // The maximum size of the data buffer
+    double sensor_to_surface_mm_; // The distance between the top of the optical sensor and the surface of the finger
+    double target_surface_offset_mm_; // The offset of the target (during calibration) from the surface of the other finger
+    std::string calibration_path_; // The file to load/save calibration from/to
+    int calib_see_through_samples_; // The number of samples to collect when determining the minimum detection threshold for the optical sensors
+    int calib_see_through_target_; // Not actually used
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/optical_weiss_finger.h:137]: Out of curiosity, why is this here? (I'm just looking at the comment right now).

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+    /**
+     * @brief cmd_response_to_msg: De-serialize sensor data from hand into a message
+     * @param resp: The response received from reading data
+     * @param resp_len: The length of the response
+     * @param devices: The number of optical sensors in the finger
+     * @param reg_lengths: The length of each register that was read
+     * @param stamp: The stamp to use for the returned message, defaults to the time at which function was called
+     * @return Message with data read from the hand
+    */
+    static  wsg_50_common::WeissFingerData cmd_response_to_msg(unsigned char* resp, unsigned int resp_len,
```
Comment (sniyaz) [inline wsg_50_driver/include/wsg_50/optical_weiss_finger.h]: We need some version of this, but that doesn't use ROS messages.

File: wsg_50_driver/src/fmf_weiss_finger.cpp [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+#include <string>
+#include <iostream>
+#include <fstream>
+
+/**
+ * @brief FMFWeissFinger: Constructor
+ * @param nh: Node handle for grabbing params
+ * @param param_prefix: Prefix for param keys
+*/
+FMFWeissFinger::FMFWeissFinger(ros::NodeHandle &nh, std::string param_prefix):
```
Comment (sniyaz) [inline wsg_50_driver/src/fmf_weiss_finger.cpp]: Ditto, param struct.

File: wsg_50_driver/src/fmf_weiss_finger.cpp [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+                                                                                initialized_(false),
```
Comment (sniyaz) [inline wsg_50_driver/src/fmf_weiss_finger.cpp:18]: Is there a good reason to initialize this here and not in the header?

File: wsg_50_driver/src/fmf_weiss_finger.cpp [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+
+    if(!nh.getParam(param_prefix+"/calib_samples", calib_samples_)) {
+        ROS_WARN("Did not receive the number of calibration samples to perform, cannot calibrate");
+        return;
+    }
+    if(calib_samples_ < 0) {
+        ROS_WARN("Number of samples must be greater than 0");
+        return;
+    }
+    if(!nh.getParam(param_prefix+"/calib_target", calib_target_)) {
```
Comment (sniyaz) [inline wsg_50_driver/src/fmf_weiss_finger.cpp]: This all will read from the param struct I think.

File: wsg_50_driver/src/fmf_weiss_finger.cpp [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+
+            // Read from the hand
+            unsigned char* resp;
+            unsigned int resp_len;
+            cmd_submit(cmd, NULL, 0, true, &resp, &resp_len);
+            cmd_lock.unlock(); // Release the lock
+
+            // Push result into message
+            if(resp_len < 2) {
+                ROS_ERROR("Did not receive response from gripper");
```
Comment (sniyaz) [inline wsg_50_driver/src/fmf_weiss_finger.cpp]: I might be wrong, but it looks like ROS here is only used for error reporting and most of this is the low level HW interface. Hopefully that means stripping out ROS should be fairly straightforward.

File: wsg_50_driver/src/hand.cpp [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+#include "wsg_50/cmd.h"
+#include "wsg_50/optical_weiss_finger.h"
+#include "wsg_50/fmf_weiss_finger.h"
+
+/**
+ * @brief Hand: Constructor
+ * @param nh: Node handle for grabbing params
+ * @param param_prefix: Prefix for param keys
+*/
+Hand::Hand(ros::NodeHandle& nh, std::string param_prefix):
```
Comment (sniyaz) [inline wsg_50_driver/src/hand.cpp]: Ditto, param struct.

File: wsg_50_driver/src/hand.cpp [hunk 14]
```diff
@@ -0,0 +1,713 @@
...
+ * @brief restart_hand_reading: Restart reading from the hand (only)
+ * @return Whether or not reading was successfully restarted
+*/
+bool Hand::restart_hand_reading() {
+  if(!initialized_) {
+    ROS_ERROR("Not properly initialized, aborting read restart");
+    return false;
+  }
+  if(!read_hand_alive_) {
+    ROS_ERROR("Hand reading never started, call start reading first, abort");
```
Comment (sniyaz) [inline wsg_50_driver/src/hand.cpp]: All of these should become prints or something like that.

File: wsg_50_driver/src/optical_weiss_finger.cpp [hunk 15]
```diff
@@ -0,0 +1,1183 @@
...
+
+// Converts a value into the corresponding hex string. Returned string will not be prefixed with '0x'
+std::string hex_val_to_str(unsigned int val);
+
+/**
+ * @brief OpticalWeissFinger: Constructor
+ * @param nh: Node handle for grabbing params
+ * @param param_prefix: Prefix for param keys
+*/
+OpticalWeissFinger::OpticalWeissFinger(ros::NodeHandle& nh, std::string param_prefix):
```
Comment (sniyaz) [inline wsg_50_driver/src/optical_weiss_finger.cpp]: Ditto, param struct.

File: wsg_50_driver/src/test_finger_read.cpp [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    hand.pause_hand_reading();
+    hand.pause_finger_reading(1);
+
+    ros::Duration(1.0).sleep();
+
+    ros::Rate loop_rate(30);
+    while(ros::ok()) {
+        wsg_50_common::WeissFingerData finger0_sample = hand.get_finger_sample(0);
+
+        unsigned int print_length = finger0_sample.data_shape.size() > 1 ? finger0_sample.data_shape[1] : finger0_sample.data_shape[0];
```
Comment (sniyaz) [inline wsg_50_driver/src/test_finger_read.cpp]: It looks like there's some non-`ros` way to read the joint states. I _think_ this is what another ROS library would expose, but we don't want to do that through ROS here.

**Review by sniyaz — CHANGES_REQUESTED** [review]
The overall code style looks awesome! I think the big changes are the `ros` things we talked about in meeting.

It looks like `ros` here isn't super-critical: just used for publishing data, error logging, and constructor params. I _think_ it should be easy to strip out all of the `ros` stuff. There are already exposed (non-`ros`) methods for sending commands to the hand and returning its state, so I think any ROS interface library we write on top of this one has everything it needs to communicate with the hand.

File: wsg_50_driver/include/wsg_50/optical_weiss_finger.h (hunk 7 above)
Comment (palanc) [reply wsg_50_driver/include/wsg_50/optical_weiss_finger.h:137]: All of the calib_*_target variables specify how far away the fingertips should be from each other when doing that particular calibration. I think I put calib_see_through_target_ in for symmetry, but it's not actually necessary for this particular calibration.

**Review by sniyaz — APPROVED** [review]

**Comment by sniyaz** [comment]
Looks solid, thanks for making the changes!
